File: examples/cloud-datafusion-functions-plugins/src/main/java/com/google/cloud/pso/common/GCPConfig.java
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.cloud.pso.common;

import com.google.cloud.ServiceOptions;
import io.cdap.cdap.api.annotation.Description;
import io.cdap.cdap.api.annotation.Macro;
import io.cdap.cdap.api.plugin.PluginConfig;
import javax.annotation.Nullable;

/**
 * Contains config properties common to all GCP plugins, like project id and service account key.
 */
public class GCPConfig extends PluginConfig {
  public static final String AUTO_DETECT = "auto-detect";

  @Description(
      "Google Cloud Project ID, which uniquely identifies a project. "
          + "It can be found on the Dashboard in the Google Cloud Platform Console.")
  @Macro
  @Nullable
  private String project;

  @Description(
      "Path on the local file system of the service account key used for authorization. Can be set"
          + " to 'auto-detect' when running on a Dataproc cluster. When running on other clusters,"
          + " the file must be present on every node in the cluster.")
  @Macro
  @Nullable
  private String serviceFilePath;

  public String getProject() {
    String projectId = project;
    if (project == null || project.isEmpty() || AUTO_DETECT.equals(project)) {
      projectId = ServiceOptions.getDefaultProjectId();
    }
    if (projectId == null) {
      throw new IllegalArgumentException(
          "Could not detect Google Cloud project id from the environment. Please specify a project"
              + " id.");
    }
    return projectId;
  }

  @Nullable
  public String getServiceAccountFilePath() {
    if (containsMacro("serviceFilePath")
        || serviceFilePath == null
        || serviceFilePath.isEmpty()
        || AUTO_DETECT.equals(serviceFilePath)) {
      return null;
    }
    return serviceFilePath;
  }
}

File: tools/adk2-graph-asset/agent/gcp_config.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
GCP / Vertex AI initialization for the ADK2 graph POC.
Mirrors the pattern used in langgraph_math_agent_v3.
"""
import os
from pathlib import Path

import vertexai
from dotenv import load_dotenv
from google.oauth2 import service_account


def init_vertex_ai(staging_bucket: str | None = None) -> dict:
    """
    Load credentials from service-account-key.json (in project root or via env),
    initialise Vertex AI, and set environment variables required by google-adk.

    Returns a dict with 'project', 'location', and 'credentials_path'.
    """
    project_root = Path(__file__).resolve().parents[1]

    # Load .env files if present
    for env_path in [project_root / ".env", project_root / "agent" / ".env"]:
        if env_path.exists():
            load_dotenv(dotenv_path=env_path, override=False)

    project_id = os.getenv("GOOGLE_CLOUD_PROJECT", "otl-eng-avstudio")
    location = os.getenv("GOOGLE_CLOUD_LOCATION", "us-central1")

    # Tell google-adk to use Vertex AI (not Google AI Studio)
    os.environ["GOOGLE_GENAI_USE_VERTEXAI"] = "1"
    os.environ["GOOGLE_CLOUD_PROJECT"] = project_id
    os.environ["GOOGLE_CLOUD_LOCATION"] = location

    # Use explicit SA key if available; otherwise fall back to ADC (used on Agent Engine).
    default_key_path = project_root / "service-account-key.json"
    key_path = Path(
        os.getenv("GOOGLE_APPLICATION_CREDENTIALS", str(default_key_path))
    ).resolve()

    init_kwargs: dict = {"project": project_id, "location": location}

    if key_path.exists():
        os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = str(key_path)
        credentials = service_account.Credentials.from_service_account_file(
            str(key_path)
        )
        init_kwargs["credentials"] = credentials
    else:
        # Running on Agent Engine or any environment with ADC configured.
        pass
    if staging_bucket:
        init_kwargs["staging_bucket"] = staging_bucket

    vertexai.init(**init_kwargs)

    return {
        "project": project_id,
        "location": location,
        "credentials_path": str(key_path),
    }
